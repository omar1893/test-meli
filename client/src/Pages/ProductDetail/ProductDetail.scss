@import "../../../src/index.scss";

.container {
  @include grid-basic($container: true);

  .detailCard {
    @include grid-positioning(2, 12);
    background-color: #ffffff;
    padding: 3.2rem;
    margin-bottom:2rem
  }

  .detailCard__main {
    @include grid-basic(10);

    .detailCard__image {
      @include grid-positioning(1, 8);
      img {
        width: 68rem;
      }

      @media screen and (max-width: 700px) {
        grid-column-end: 11;
        img {
          width: 100%;
        }
    }
    }

    .detailCard__info {
      @include grid-positioning(8, 11);
      .detailCard__info--condition {
        font-size: 1.4rem;
        margin-bottom: 1.6rem;
      }
      .detailCard__info--title {
        font-size: 2.4rem;
        margin-bottom: 3.2rem;
      }
      .detailCard__info--price {
        font-size: 4.6rem;
        margin-bottom: 3.2rem;
      }
      .detailCard__info--button {
        background-color: $blue-cta;
        display: flex;
        justify-content: center;
        align-items: center;
        border: none;
        border-radius: 4px;
        color: #ffffff;
        padding-top: 2rem;
        padding-bottom: 2rem;
        width: 100%;
        font-size: 2rem;
      }

      @media screen and (max-width: 700px) {
        grid-column-start: 1;
        text-align: center;
    }
    }
  }

  .detailCard__description {
    @include grid-basic(10);

    .detailCard__description--title {
      @include grid-positioning(1, 12);
      @include grid-positioning-rows(1, auto);
      font-size: 2.8rem;
      margin-bottom: 3.2rem;
    }

    .detailCard__description--text {
      @include grid-positioning-rows(2, auto);
      @include grid-positioning(1, 8);
      font-size: 1.6rem;
      color: $dark-gray;

      @media screen and (max-width: 700px) {
        grid-column-end: 11;
    }
    }
  }
}
