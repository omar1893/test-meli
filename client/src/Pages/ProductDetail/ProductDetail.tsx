import React, { useEffect, useState } from "react";
import axios from "axios";
import { useParams } from "react-router-dom";
import { ItemInfo } from "../../interfaces";
import "./ProductDetail.scss";
import { priceFormatter } from "../../utils/priceFormatter";
import { BreadCrumbs } from "../../components/Breadcrumbs";
import { Loading } from "../../components/Loading";
import { CONDITION } from "../../utils/condition";

const ProductDetail = () => {
  const [item, setItem] = useState<ItemInfo>();
  const [categories, setCategories] = useState<string[]>([]) 
  const { id } = useParams();
  const [loading, setLoading] = useState(false)

  const getItem = async () => {
    if (id) {
      setLoading(true)
      try {
        const response = await axios.get(`http://localhost:3001/api/items/${id}`);
        const { data } = await response;
        setItem(data.item);
        setLoading(false)
      } catch (error) {
        setLoading(false)
        console.log(error)
      }
    }
  };

  useEffect(() => {
    const lsCategories = localStorage.getItem('categories');
    if (lsCategories) {
      setCategories(JSON.parse(lsCategories));
    }
    getItem();
  }, [id]);

  return loading ? (<Loading/>) : (
    <div className="container">
      <BreadCrumbs categories={categories}/>
      <div className="detailCard">
        <div className="detailCard__main">
          <div className="detailCard__image">
            <img src={item?.picture} alt="" />
          </div>
          <div className="detailCard__info">
            <p className="detailCard__info--condition">
              {CONDITION[item?.condition === 'new' ? 'new' : 'old']} - {item?.sold_quantity} vendidos
            </p>
            <p className="detailCard__info--title">{item?.title}</p>
            <p className="detailCard__info--price">
              {item && priceFormatter(item.price.amount.toString(), item?.price.currency)} 
            </p>
            <button className="detailCard__info--button">Comprar</button>
          </div>
        </div>
        <div className="detailCard__description">
          <p className="detailCard__description--title">
            Descripci√≥n del producto
          </p>
          <p className="detailCard__description--text">{item?.description}</p>
        </div>
      </div>
    </div>
  );
};

export default ProductDetail;
