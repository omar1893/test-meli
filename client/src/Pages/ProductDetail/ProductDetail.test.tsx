import React from "react";
import { render, screen } from "@testing-library/react";
import axios from "axios";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import ProductDetail from "./ProductDetail";

jest.mock("axios");

describe("ProductDetail", () => {
  beforeEach(() => {
    axios.get = jest.fn().mockResolvedValueOnce({
      data: {
        item: {
          id: "1",
          picture: "image-url",
          condition: "New",
          sold_quantity: 10,
          title: "Product Title",
          price: {
            amount: 99.99,
            currency: "USD",
            decimals: 0
          },
          description: "Product Description",
        },
      },
    });
  });

  it("Should match snapshot", async () => {
    const { container } = render(
      <MemoryRouter initialEntries={["/items/1"]}>
        <Routes>
          <Route path="/items/:id" element={<ProductDetail />} />
        </Routes>
      </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
  });

  it("Should display product information", async () => {
    render(
      <MemoryRouter initialEntries={["/items/1"]}>
        <Routes>
          <Route path="/items/:id" element={<ProductDetail />} />
        </Routes>
      </MemoryRouter>
    );
    expect(await screen.findByText("Product Title")).toBeInTheDocument();
    expect(await screen.findByText("99,99 US$")).toBeInTheDocument();
    expect(await screen.findByText("Product Description")).toBeInTheDocument();
  });
});
