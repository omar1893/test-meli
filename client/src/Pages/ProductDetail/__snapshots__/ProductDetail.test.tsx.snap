// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetail Should match snapshot 1`] = `
<div>
  <div>
    <p
      class="loading-message"
    >
      Cargando...
    </p>
  </div>
</div>
`;
